rust/lance/src/dataset.rs

```diff
@@ -822,70 +822,134 @@ impl Dataset {
-            .ok_or_else(|| Error::Internal {
-                message: "ROW_ID column is not UInt64Array".into(),
-            })?
-            .values();
-
-        let remapping_index: UInt64Array = row_ids
-            .iter()
-            .filter_map(|o| returned_row_ids.binary_search(o).ok().map(|pos| pos as u64))
-            .collect();
+        if row_id_meta.contiguous {
+            // Fastest path: Can use `read_range` directly
+            let start = row_ids.first().expect("empty range passed to take_rows");
+            let fragment_id = (start >> 32) as usize;
+            let range_start = *start as u32 as usize;
+            let range_end =
+                *row_ids.last().expect("empty range passed to take_rows") as u32 as usize;
+            let range = range_start..(range_end + 1);
 
-        // Remove the row id column.
-        let keep_indices = (0..one_batch.num_columns() - 1).collect::<Vec<_>>();
-        let one_batch = one_batch.project(&keep_indices)?;
-        let struct_arr: StructArray = one_batch.into();
-        let reordered = take(&struct_arr, &remapping_index, None)?;
-        Ok(as_struct_array(&reordered).into())
+            let fragment = self.get_fragment(fragment_id).ok_or_else(|| {
+                Error::invalid_input(format!("row_id belongs to non-existant fragment: {start}"))
+            })?;
+
+            let reader = fragment.open(projection).await?;
+            reader.read_range(range).await
+        } else if row_id_meta.sorted {
+            // Don't need to re-arrange data, just concatenate
+
+            let mut batches: Vec<RecordBatch> = Vec::new();
+            let mut current_fragment = row_ids[0] >> 32;
+            let mut current_start = 0;
+            let mut row_ids_iter = row_ids.iter().enumerate();
+            'outer: loop {
+                let (fragment_id, range) = loop {
+                    if let Some((i, row_id)) = row_ids_iter.next() {
+                        let fragment_id = row_id >> 32;
+                        if fragment_id != current_fragment {
+                            let next = (current_fragment, current_start..i);
+                            current_fragment = fragment_id;
+                            current_start = i;
+                            break next;
+                        }
+                    } else if current_start != row_ids.len() {
+                        let next = (current_fragment, current_start..row_ids.len());
+                        current_start = row_ids.len();
+                        break next;
+                    } else {
+                        break 'outer;
+                    }
+                };
+
+                let fragment = self.get_fragment(fragment_id as usize).ok_or_else(|| {
+                    Error::invalid_input(format!(
+                        "row_id belongs to non-existant fragment: {}",
+                        row_ids[current_start]
+                    ))
+                })?;
```

Comment:
This is a potential opportunity for threading in the future.  You could launch this as a new task while you continue to gather fragment / row ids on the current thread.

Reply:
👍 It would also be nice if `get_fragment()` wasn't implemented as a linear scan.